app/services/foreman/provision/ssh.rb

```diff
@@ -50,64 +71,90 @@ def command_prefix
-      Timeout.timeout(8) do
-        ssh.run('pwd')
-      end
-    rescue Errno::ECONNREFUSED
-      logger.debug "Connection refused for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Errno::EHOSTUNREACH
-      logger.debug "Host unreachable for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::Disconnect
-      logger.debug "Host dropping connections for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::ConnectionTimeout
-      logger.debug "Host timed out for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::AuthenticationFailed
-      logger.debug "Auth failed for #{username} at #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Timeout::Error
-      retry
-    rescue => e
-      Foreman::Logging.exception("SSH error", e)
-    end
+  def scp
+    logger.debug "Uploading template to remote system '#{remote_host}'"
+    logger.debug scp_cmd
+    _stdout, stderr, status = Open3.capture3(*scp_cmd)
+
+    return if status.success?
+
+    logger.error stderr
+    raise Foreman::Exception.new("Failed to upload '#{@template}' to remote system at #{address}")
   end
 
-  def ssh
-    Fog::SSH.new(address, username, options.merge(:timeout => 4))
+  def passwd_cmd(mask_passwd: false)
+    return if @options[:password].empty?
+
+    value = mask_passwd ? '*****' : @options[:password]
+    ["sshpass", "-p", value]
   end
 
-  def scp
-    Fog::SCP.new(address, username, options)
+  def ssh_key
+    return if @options[:key_data].empty?
+    ["-i", @tmp_key_path]
+  end
+
+  def ssh_cmd(host_command, mask_passwd: false)
+    [passwd_cmd, "ssh", ssh_key, ssh_options, remote_host, host_command].flatten.compact
+  end
+
+  def scp_cmd
+    [passwd_cmd, "scp", ssh_key, ssh_options, @template, "#{remote_host}:~/#{remote_script}"].flatten.compact
+  end
+
+  def ssh_options
+    opts = []
+
+    opts << "-o ConnectTimeout=6"
+    opts << "-o Compression=yes"
+    opts << "-o ControlMaster=auto"
+    opts << "-o ControlPath=#{socket_file}"
+    opts << "-o ControlPersist=yes"
+
+    methods = []
+    methods << 'publickey' if @options[:key_data].present?
+    methods << 'password' if @options[:password].present?
+
+    opts << "-o PreferredAuthentications=#{methods.join(',')}"
+
+    opts
+  end
+
+  def create_tmp_key
+    return if @options[:key_data].empty?
+
+    file = Tempfile.new("ssh-key-#{uuid}")
+    file.write(@options[:key_data])
+    file.close
+    file.path
+  end
+
+  def check_auth_options
+    return if @options[:key_data].present? || @options[:password].present?
+
+    raise Foreman::Exception.new("Missing :key_data or :password options, no authentication method available")
+  end
+
+  def socket_file
+    file = Tempfile.new("ssh-socket-#{@uuid}")
+    file.close
+    file.path
+  end
+
+  def cleanup_files
+    File.delete(@tmp_key_path) if @options[:key_data].present?
+    File.delete(socket_file)
```

Comment:
This can raise exceptions, like `Errno::ENOENT` if it doesn't exist. See https://rubyapi.org/o/file#method-c-delete. I'd suggest:
```suggestion
    files = [socket_file]
    files << @tmp_key_path if @options[:key_data].present?
    FileUtils.rm_f(files)
```